pkg/provisioners/managers/virtualcluster/provisioner.go

```diff
@@ -0,0 +1,300 @@
+	regionapi "github.com/unikorn-cloud/region/pkg/openapi"
+
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrLabel = errors.New("required label missing")
+
+	ErrAnnotation = errors.New("required annotation missing")
+
+	ErrResourceDependency = errors.New("resource dependency error")
+)
+
+type ApplicationReferenceGetter struct {
+	cluster *unikornv1.VirtualKubernetesCluster
+}
+
+func newApplicationReferenceGetter(cluster *unikornv1.VirtualKubernetesCluster) *ApplicationReferenceGetter {
+	return &ApplicationReferenceGetter{
+		cluster: cluster,
+	}
+}
+
+func (a *ApplicationReferenceGetter) getApplication(ctx context.Context, name string) (*unikornv1core.HelmApplication, *unikornv1core.SemanticVersion, error) {
+	namespace, err := coreclient.NamespaceFromContext(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// TODO: we could cache this, it's from a cache anyway, so quite cheap...
+	cli, err := coreclient.ProvisionerClientFromContext(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	key := client.ObjectKey{
+		Namespace: namespace,
+		Name:      *a.cluster.Spec.ApplicationBundle,
+	}
+
+	bundle := &unikornv1.VirtualKubernetesClusterApplicationBundle{}
+
+	if err := cli.Get(ctx, key, bundle); err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := bundle.Spec.GetApplication(name)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	key = client.ObjectKey{
+		Namespace: namespace,
+		Name:      *reference.Name,
+	}
+
+	application := &unikornv1core.HelmApplication{}
+
+	if err := cli.Get(ctx, key, application); err != nil {
+		return nil, nil, err
+	}
+
+	return application, &reference.Version, nil
+}
+
+func (a *ApplicationReferenceGetter) vCluster(ctx context.Context) (*unikornv1core.HelmApplication, *unikornv1core.SemanticVersion, error) {
+	return a.getApplication(ctx, "vcluster")
+}
+
+// Options allows access to CLI options in the provisioner.
+type Options struct {
+	// identityOptions allow the identity host and CA to be set.
+	identityOptions *identityclient.Options
+	// regionOptions allows the region host and CA to be set.
+	regionOptions *regionclient.Options
+	// clientOptions give access to client certificate information as
+	// we need to talk to identity to get a token, and then to region
+	// to ensure cloud identities and networks are provisioned, as well
+	// as deptovisioning them.
+	clientOptions coreclient.HTTPClientOptions
+	// domain vclusters should appear in.
+	domain string
+}
+
+func (o *Options) AddFlags(f *pflag.FlagSet) {
+	if o.identityOptions == nil {
+		o.identityOptions = identityclient.NewOptions()
+	}
+
+	if o.regionOptions == nil {
+		o.regionOptions = regionclient.NewOptions()
+	}
+
+	o.identityOptions.AddFlags(f)
+	o.regionOptions.AddFlags(f)
+	o.clientOptions.AddFlags(f)
+
+	f.StringVar(&o.domain, "vcluster-domain", "dev.spjmurray.co.uk", "DNS domain for vclusters to be hosts of.")
```

Comment:
You might want to change this to `example.com` before merging, just to make it clear it 1. is arbitrary and 2. is a value that has to be supplied.